### ISSUE TITLE
How to mask or ignore target features during model training

### ISSUE THREAD

**Issue by yunakkano** [unit8co/darts#2381]
I am having a difficulty to train TSMixer model in my current project where target values wouldn't be available at the time of prediction. 

Is there any option to mask or ignore past target values during model traing?

**Comment by dennisbader** [comment]
Can you elaborate on what you mean by "the target values are not available at prediction time"?

Just for clarification:
- the target variables are the ones that you want to forecast (and not covariates/external features that you do not want to forecast)
- for training our torch models (neural networks), you need at least `input_chunk_lenght+output_chunk_length+output_chunk_shift` points per target series (for both single series or multiple / a list of series). There are two options for too short series:
  - If you work with multiple series, you could just drop the too short ones from the input `series`.
  - For some series (e.g. sales data) prepending some values before the start of the series can help. Let's say you have sales data for a product, but the product has been launched recently, and the sales history is not long enough yet. Here you could prepend zeros (zero sales) until the length covers the required model input time frame. Additionally, you could add a new (binary) covariate feature that flags these time periods (e.g. value `1` if the time is before the launch and `0` otherwise)

**Comment by yunakkano** [comment]
Hi @madtoinou and @dennisbader ,

Thank you both for your helpful responses. I appreciate your insights and suggestions.

To clarify my scenario, I have a timeseries dataset with feature columns x1, x2, x3, and y. The variable y is the target I want to predict, but it is only available after x1, x2, and x3 are observed. This means that y cannot be used during the training of the prediction model.

Assuming a time horizon t_0, t_1, ..., t_now, t_now+1, and so on, we are currently at t_now. I want to train the model with the features x1, x2, and x3 from the time section t_0 to t_now in order to predict y values from t_now+1 onwards.

To address your question, @dennisbader , regarding what I mean by "the target values are not available at prediction time," here is some pseudo code to illustrate my situation. I aim to build a `TSMixerModel` that takes 6 past time steps to predict 6 future time steps, but I'm unsure how to set the 'target' variable. If I include y in the 'target', the model will be trained with y values, which is not applicable in my scenario. On the other hand, if I exclude y from the 'target', the model will not be trained to predict y values. Here is the pseudo code for reference:

`model = TSMixerModel(
    input_chunk_length=6,
    output_chunk_length=6
)
model.fit(target, past_covariates=past_cov)`

Given this context, could you provide more specific advice on how to handle this scenario within the Darts framework? Is there a way to mask or ignore the y values during model training while still training the model to predict y values from t_now+1 onwards?

Thank you again for your assistance.

Best regards,

**Comment by dennisbader** [comment]
Hi @yunakkano let me see if I understand your problem correctly. If you were to predict in this moment, your scenario would look something like below?

| var\time | t0 | t1 | t2 | t3 | t4 | tnow | tpred1 | tpred... | tpred6 |
|----------|----|----|----|----|----|------|--------|----------|--------|
| y        | 1  | 2  | 3  | 4  | 5  | -    | ?      | ?        | ?      |
| x1       | 5  | 6  | 7  | 8  | 9  | 10   | -      | -        | -      |


- `y` is your target variable
- `x1` is one of your covariate features
- `tpred` is the current time step after which you want to predict `tpred1` until `tpred6` (horizon=output_chunk_length=6 points)
- `t0` until `tnow` is you input window (`input_chunk_length=6` points)
- `?` are the `y` values you want to predict
- `-` means the variable is not available at the given time step

There are two ways to this.

- with our neural networks (such as `TSMixer`), it can be done but you will not be able to use `tnow` of `x1, x2, ...`. For this you can use `output_chunk_shift=1` at model creation. This will use `t0` until `t4` (in other words `tnow-1`) from `y` and `x*` to predict `tpred1` (in other words `tnow+1`).

  ```
  model = TSMixerModel(
      input_chunk_length=6,
      output_chunk_length=6,
      output_chunk_shift=1,
  )
  model.fit(target, past_covariates=past_cov)
  ```
- with our regression models (any of them) it can be done using all available information. The configuration below will use all the information from the table. In this case we need to use the covariates as `future_covariates`. When predicting, make sure that your target series ends at `tnow-1`, and the covariates series at `tnow`.

  ```
  model = RegressionModel(
      lags=5,
      lags_future_covariates=[-6, -5, -4, -3, -2, -1],
      output_chunk_length=6,
      output_chunk_shift=1,
  )
  model.fit(target, future_covariates=past_cov)
  ```

**Comment by dennisbader** [comment]
Hi @yunakkano, in that case, the only option is to use our regression models without using any information of the target series `y` as input (with `lags=None`).

```
model = RegressionModel(
    lags=None,
    lags_past_covariates=6,
    output_chunk_length=6
)
model.fit(target, past_covariates=past_cov)
```

**Comment by yunakkano** [comment]
Hi @dennisbader,

Thank you for the recommendation to use the regression models without any information from the target series y as input.

I have a somewhat naive question: would it be possible to customize the original TSMixer model to handle my scenario by, for example, fixing y to a specific value during training to prevent the model from learning from the target?

Specifically, do you think a customization where we replace all information of the target column with zero in the forward() method of the _TSMixerModule class would be effective?

Thank you again for your assistance.

**Comment by yunakkano** [comment]
Wow, @dennisbader ! I never imagined you could come up with such an amazing idea.

Thank you so much for this detailed explanation and the code example. I'm really excited to try this out.

Just to confirm, based on your implementation here:

x_past = x_past[:, :, self.input_target_dim:]
the target feature y must be located at the beginning of the tensor, right?

Thanks again for your incredible support.

**Comment by dennisbader** [comment]
Hi @yunakkano, no worries.

Yes, they must be located at the beginning. This is handled automatically by our datasets, so you shouldn't have to worry about it.

**Comment by tRosenflanz** [comment]
I wish this issue was created 3 days ago when I wanted to do exactly the same. 
In my case I didn't want target and future covariates to be included so I just override them with 0s 
```
from typing import Tuple
from darts.utils.data.inference_dataset import MixedCovariatesInferenceDataset
from darts.utils.data.sequential_dataset import MixedCovariatesSequentialDataset
from numpy import ndarray
# past_target,
# past_covariates,
# historic_future_covariates,
# future_covariates,
# static_covariates,
# future_target
### remove the past target and historic future covariates because they get concatenated into model inputs
class StrippedMixedCovariatesSequentialDataset(MixedCovariatesSequentialDataset):
    def __getitem__(self, idx) -> Tuple[ndarray, ndarray | None, ndarray | None, ndarray | None, ndarray | None, ndarray]:
        vals = list(super().__getitem__(idx))
        vals[0] = vals[0]*0
        vals[2] = vals[2]*0
        # vals[0] = vals[1][:,[1]]
        # vals[2] = vals[1][:,[0]]
        return tuple(vals)
        
class StrippedMixedCovariatesInferenceDataset(MixedCovariatesInferenceDataset):
    def __getitem__(self, idx) -> Tuple[ndarray, ndarray | None, ndarray | None, ndarray | None, ndarray | None, ndarray]:
        vals = list(super().__getitem__(idx))
        vals[0] = vals[0]*0
        vals[2] = vals[2]*0
        # vals[0] = vals[1][:,[1]]
        # vals[2] = vals[1][:,[0]]
        return tuple(vals)
```

**Comment by dennisbader** [comment]
Hi @tRosenflanz, this is certainly another way to do it, just that in this case your model complexity is larger for handling the redundant information.

**Comment by SaltedfishLZX** [comment]
This discussion is really helpful. hope all models will have the option to ignore past target values

**Comment by madtoinou** [comment]
Closing this issue, the feature request is covered by #2151 (and the code snippet shared are probably a good start to implement such a thing).
